### DISCUSSION TITLE
Willing to bring improvements to LinearMouse

### DISCUSSION THREAD

**Discussion by bfmatei** [linearmouse/linearmouse#736] [Ideas]
Hello!

Let me start by saying congrats for the codebase! I have no prior Swift experience but I managed to pick it up on the go thanks to the overall architecture of LinearMouse.

Last week I forked the repo and I started working on a couple of improvements. I am willing to contribute them back to this repo if you want but there are a couple of gotchas that we need to discuss before.

## [Mice that emit non-standard events](https://github.com/linearmouse/linearmouse/issues/708)

Some mice don't emit standard events for the additional buttons but they do have on-board memory and you can bind those buttons in their native software. SteerMouse circumvents this by simply allowing you to bind letters A-Z and F1-F6 to the mouse buttons and then translating those to Button 1 - Button 32.

In order to achieve this, there are a couple of big changes that need to be brought, mainly around `PointerKit`.

### Replace `PointerKit` with a more generic `InputKit` that also looks for keyboard and keypad devices.

This is needed in order to catch the keyboard events and then properly extract the input device from the event as some devices are acting pretty weird.

Let's take for example Roccat Kone XP Air, although I found that others act similarly:
   - on bluetooth and directly connected via cable it provides one single device with all the necessary usage pages and usages. The device has the same vendor ID and product ID for both types of connections. These types of devices don't need any special treatment.
   - the 2.4GHz dongle provides 3 devices with 3 different usage pages and usages (one for the mouse, one for the keyboard and one auxiliary for the lights). These device all have the same vendor ID and product ID and they are different from the ones provided by bluetooth and direct USB connection. These devices need to have a "primary input device" (that being the mouse/pointer device) and all the events emitted to the other devices will be allocated to this primary device.

### Catch and manipulate the key up/down events

This can be done either via an event transformer or via event manipulation in input value callback. I preferred the first approach as it's a little bit easier to work with instead of having a huge switch for each button but this can also increase the latency a little bit (although I personally didn't notice it tho').

### Events transformers should also listen to mouse move events

This isn't a huge change from a code perspective but it has the potential to impact the performance.

This is needed in order to properly emit mouse dragged events

## [Keyboard mappings](https://github.com/linearmouse/linearmouse/discussions/451)

Because of the `PointerKit` -> `InputKit` change, we can easily facilitate keyboard mappings, bringing LinearMouse more closer to other software like Better Mouse and SteerMouse.

This can be a separate section in the menu, similarly to the "Buttons" one that already exist.

## [Gesture mappings](https://github.com/linearmouse/linearmouse/issues/729)

Some other software have support for gestures. For example, holding button 4 and dragging the mouse up can bring up mission control. I managed to implement this via transformers without any problems although I am not convinced where to place this feature: allowing this kind of inputs in the button mapping section or creating a separate section in the menu. I would go for the former though.

## Dock gestures

Right now, LinearMouse support magnify events and navigation events (aka 2 fingers swiping). In addition to this, other software also support dock gestures (aka 3 fingers swiping).

Implementing this is extremely complicated for the time being as opposed to the other types of gestures, these ones are not that easy to work with. I am doing some reverse engineering in order to be able to emit these events but it will take some time.

## [Smooth scrolling](https://github.com/linearmouse/linearmouse/issues/730)

I can implement smooth scrolling as well.

## [Disable device](https://github.com/linearmouse/linearmouse/discussions/636)

I can implement this feature as well.

## Consensus 

Now that I presented all the features that I am willing to work on (or already started to work on for some of them), what do you think about them? Are they something you want in LinearMouse? If yes, I can work on bringing them to LinearMouse but I need to have a guarantee that this product will stay open source and it won't take a freemium or paid approach.

**Comment by lujjjh** [comment]
Thank you for your tremendous work! Enhancements to LinearMouse are always welcome. Due to a packed schedule recently, my replies might not be as prompt as usual. But I will certainly make time to try out your fork.

> Some mice don't emit standard events for the additional buttons but they do have on-board memory and you can bind those buttons in their native software. SteerMouse circumvents this by simply allowing you to bind letters A-Z and F1-F6 to the mouse buttons and then translating those to Button 1 - Button 32.

I'm trying to understand the process here. Do I first assign the additional mouse buttons to a letter or one of the F1-F6 keys using the mouse's native software (such as Logi Options+), and then map those keys to a specific action in LinearMouse / SteerMouse?

> This can be done either via an event transformer or via event manipulation in input value callback. I preferred the first approach as it's a little bit easier to work with instead of having a huge switch for each button but this can also increase the latency a little bit (although I personally didn't notice it tho').

Yeah, event transformers are specifically designed for such purposes. The slight increase in latency is typically imperceptible. The first approach offers a more straightforward implementation and are easier to maintain compared to managing a complex switch-case for each button in input value callbacks.

> ### Events transformers should also listen to mouse move events
> This isn't a huge change from a code perspective but it has the potential to impact the performance.
> This is needed in order to properly emit mouse dragged events

I share the concerns regarding performance since I've experienced a noticeable increase in CPU usage when listening to mouse move events, which could potentially displease users who don't require this feature.

However, there's room for optimization, such as initiating the listening process only when related features are enabled, or when a mouse button is pressed.

And there are events like `CGEventType.xxxMouseDragged`. I'm not sure if they help.

*(to be continued...)*

> I need to have a guarantee that this product will stay open source and it won't take a freemium or paid approach.

Certainly, LinearMouse will remain open source and free of charge. [The donations received](https://opencollective.com/linearmouse) thus far are sufficient to cover essential costs such as certificates and server expenses for LinearMouse itself.

**Reply by bfmatei** [reply]
> Certainly, LinearMouse will remain open source and free of charge. [The donations received](https://opencollective.com/linearmouse) thus far are sufficient to cover essential costs such as certificates and server expenses for LinearMouse itself.

Perfect <3

> I'm trying to understand the process here. Do I first assign the additional mouse buttons to a letter or one of the F1-F6 keys using the mouse's native software (such as Logi Options+), and then map those keys to a specific action in LinearMouse / SteerMouse?

Yup, that's the exact process that I was thinking of and AFAIK SteerMouse works the same. You can see it in the `Does SteerMouse support Razer mice?` section of their [FAQ](https://plentycom.jp/en/steermouse/support.html).

> I share the concerns regarding performance since I've experienced a noticeable increase in CPU usage when listening to mouse move events, which could potentially displease users who don't require this feature.

At least on M3 Pro, it is negligible. Moving the mouse continuously while also typing takes the CPU to 2-3% tops. But yeah, I'll make some improvements to the process as well.

> And there are events like CGEventType.xxxMouseDragged. I'm not sure if they help.

That's exactly why we need the `mouseDragged` event: be able to not only map to clicks (emitting `keyDown -> mouseDown` and then `keyUp -> xxMouseUp`) but also emit dragging events (emit `keyDown -> xxMouseDown`, then if mouse is moved `mouseMoved -> xxMouseDragged` and finally `keyUp -> xxMouseUp`). This isn't a must though but it's a nice feature that allows things like gestures with additional buttons.

I'm currently working on separating everything in atomic commits so you can better look over the code step by step.
